### PR TITLE
Show a better error when someone throws undefined

### LINKED ISSUE #3186: Throw undefined not handled correctly
- [x] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Expected Behavior

Throwing undefined in the `renderToHTML` shouldn't halt the request.

## Current Behavior

Throwing undefined halt the request.

## Steps to Reproduce (for bugs)

1. Throw `undefined` in a `.getInitialProps()` method.

## Context

To be more specific an error is raised from this line:
https://github.com/zeit/next.js/blob/2ec397b3564d5dbca933e8def9cd0801ebd3a2b5/server/index.js#L321

> unhandledRejection reason TypeError: Cannot read property 'code' of undefined
     at Server._callee18$ (/app/node_modules/next/dist/server/index.js:943:37)

Yes, throwing undefined is [supported by JavaScript](https://stackoverflow.com/questions/30049093/javascript-is-it-safe-to-throw-undefined). It was a mistake on my end. I don't throw undefined on purpose 😁 .

## Your Environment

| Tech    | Version |
|---------|---------|
| next    |  v4.0.2       |
| node    | v8.5.0        |
| OS      | Linux        |
| browser | Chrome        |

### PR DESCRIPTION
**Pull request by ijjk**
Fixes: #3186

### CODE REVIEW

File: packages/next/server/next-dev-server.js [hunk 1]
```diff
@@ -120,6 +120,10 @@ export default class DevServer extends Server {
       return super.renderErrorToHTML(compilationErr, req, res, pathname, query)
     }
 
+    if (!err) {
+      err = new Error('Something threw undefined somewhere during render...')
```
Comment (dav-is) [inline packages/next/server/next-dev-server.js]:
```suggestion
      err = new Error('An undefined error was thrown sometime during render...')
```

File: test/integration/basic/pages/throw-undefined.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+export default function ThrowUndefined (props) {
+  return (
+    <div>throw-undefined</div>
+  )
+}
+
+ThrowUndefined.getInitialProps = () => {
```
Comment (dav-is) [inline test/integration/basic/pages/throw-undefined.js]: We should export at the bottom after this

File: test/integration/basic/test/rendering.js [hunk 3]
```diff
@@ -289,5 +289,10 @@ export default function ({ app }, suiteName, render, fetch) {
         expect($('#asPath').text()).toBe('Current asPath: /nav/with-hoc')
       })
     })
+
+    it('should show better error when undefined is thrown', async () => {
```
Comment (dav-is) [inline test/integration/basic/test/rendering.js]: "better" doesn't have any context outside of this PR

File: test/integration/basic/test/rendering.js [hunk 4]
```diff
@@ -289,5 +289,10 @@ export default function ({ app }, suiteName, render, fetch) {
...
+      const $ = await get$('/throw-undefined')
+      expect($('body').text()).toMatch('Something threw undefined somewhere during render...')
```
Comment (dav-is) [inline test/integration/basic/test/rendering.js]:
```suggestion
      expect($('body').text()).toMatch('An undefined error was thrown sometime during render...')
```

File: test/integration/basic/pages/throw-undefined.js (hunk 2 above)
Comment (ijjk) [reply test/integration/basic/pages/throw-undefined.js]: Updated

File: test/integration/basic/test/rendering.js (hunk 3 above)
Comment (ijjk) [reply test/integration/basic/test/rendering.js]: Good point, updated

**Review by dav-is — APPROVED** [review]
🙌🏼

File: packages/next/server/next-dev-server.js [hunk 5]
```diff
@@ -120,6 +120,10 @@ export default class DevServer extends Server {
       return super.renderErrorToHTML(compilationErr, req, res, pathname, query)
     }
 
+    if (!err && res.statusCode === 500) {
+      err = new Error('An undefined error was thrown sometime during render...')
```
Comment (timneutkens) [inline packages/next/server/next-dev-server.js]: The err.sh link is missing here

**Pull request merged**